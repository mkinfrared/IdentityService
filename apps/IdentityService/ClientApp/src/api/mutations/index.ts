export * from "./loginMutation";

export * from "./registerMutation";

export * from "./verifyEmailMutation";

export * from "./submitConsent";
