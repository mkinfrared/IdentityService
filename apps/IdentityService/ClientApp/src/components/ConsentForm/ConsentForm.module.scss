@import "../../styles/common";

.ConsentForm {
  @extend %loginRegisterForm;

  .description {
    margin-top: 5px;
  }
}

.Story {
}
