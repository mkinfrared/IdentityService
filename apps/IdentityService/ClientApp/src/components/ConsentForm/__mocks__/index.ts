export { default } from "./ConsentForm.mock";
