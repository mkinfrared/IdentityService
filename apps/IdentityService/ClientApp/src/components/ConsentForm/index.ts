export { default } from "./ConsentForm.container";

export * from "./ConsentForm.type";
