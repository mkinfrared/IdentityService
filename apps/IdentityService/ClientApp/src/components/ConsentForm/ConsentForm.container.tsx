import { memo } from "react";

import { ApiScopeUpdateDto, ConsentUpdateDto } from "api/types";
import { useConsentQuery } from "hooks/useConsentQuery";

import ConsentFormFC from "./ConsentForm";
import { ConsentFormContainerProps } from "./ConsentForm.type";

const ConsentFormContainer = ({ returnUrl }: ConsentFormContainerProps) => {
  const { data } = useConsentQuery(returnUrl);

  const mapScope = (scope: ApiScopeUpdateDto) => ({
    ...scope,
    isPermitted: scope.required,
  });

  const defaultValues: ConsentUpdateDto = {
    rememberConsent: false,
    description: "",
    permissionGranted: false,
    redirectUrl: data?.returnUrl ?? "",
    identityScopes: data?.identityScopes?.map(mapScope) ?? [],
    apiScopes: data?.apiScopes?.map(mapScope) ?? [],
  };

  return (
    <ConsentFormFC
      allowRememberConsent={!!data?.allowRememberConsent}
      clientName={data?.clientName ?? ""}
      defaultValues={defaultValues}
      returnUrl={returnUrl}
    />
  );
};

export default memo(ConsentFormContainer);
