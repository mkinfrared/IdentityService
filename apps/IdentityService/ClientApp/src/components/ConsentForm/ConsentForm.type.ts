import { ConsentUpdateDto } from "api/types";

export type ConsentFormProps = {
  allowRememberConsent: boolean;
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  clientName: string;
  defaultValues: ConsentUpdateDto;
  returnUrl: string;
};

export type ConsentFormContainerProps = Pick<ConsentFormProps, "className">;
