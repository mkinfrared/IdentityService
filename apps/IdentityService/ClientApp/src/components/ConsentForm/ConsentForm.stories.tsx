import { Meta, Story } from "@storybook/react";

import { consentReadDtoMock } from "api/queries/getConsent/__mocks__";
import { withQueryProvider } from "utils/storybookDecorators";

import { ConsentForm } from "./ConsentForm";
import { ConsentFormProps } from "./ConsentForm.type";
import css from "./Story.module.scss";

export default {
  title: "COMPONENTS/ConsentForm",
  component: ConsentForm,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
  decorators: [withQueryProvider],
} as Meta;

const Template: Story<ConsentFormProps> = (args) => (
  <div className={css.Story}>
    <ConsentForm {...args} />
  </div>
);

const Default = Template.bind({});

Default.args = {
  defaultValues: {
    rememberConsent: false,
    apiScopes: consentReadDtoMock.apiScopes.map((scope) => ({
      ...scope,
      description: scope.description ?? "",
      isPermitted: !!scope.required,
    })),
    identityScopes: consentReadDtoMock.identityScopes.map((scope) => ({
      ...scope,
      description: scope.description ?? "",
      isPermitted: false,
    })),
    description: "",
    permissionGranted: false,
    redirectUrl: "/foo/bar",
  },
  clientName: "foobar",
};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
