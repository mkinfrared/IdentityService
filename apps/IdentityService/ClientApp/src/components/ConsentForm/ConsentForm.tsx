import { classNames } from "@identity-service/core";
import { Button, Fieldset, Heading } from "@identity-service/ui";
import { Fragment, memo, useCallback, useMemo } from "react";
import { useFieldArray, useForm } from "react-hook-form";

import CheckboxField from "components/CheckboxField";
import ToggleField from "components/ToggleField";
import { useConsentSubmitMutation } from "hooks/useConsentSubmitMutation";

import css from "./ConsentForm.module.scss";
import { ConsentFormProps } from "./ConsentForm.type";

const ConsentForm = ({
  allowRememberConsent,
  className,
  clientName,
  defaultValues,
}: ConsentFormProps) => {
  const { isLoading, mutateAsync } = useConsentSubmitMutation();

  const { getValues, register, control } = useForm({
    defaultValues,
  });

  const { fields: identityFields } = useFieldArray({
    control,
    name: "identityScopes",
  });

  const { fields: apiFields } = useFieldArray({
    control,
    name: "apiScopes",
  });

  const identityScopes = useMemo(
    () =>
      identityFields.map((field, index) => (
        <Fragment key={field.id}>
          <ToggleField
            {...register(`identityScopes.${index}.isPermitted` as const)}
            label={field.displayName ?? ""}
            className={css.input}
            defaultChecked={field.required}
            readOnly={field.required}
          />
          <Heading variant="h6" className={css.description}>
            {field.description}
          </Heading>
        </Fragment>
      )),
    [identityFields, register],
  );

  const apiScopes = useMemo(
    () =>
      apiFields.map(({ displayName, description, required, id }, index) => (
        <Fragment key={id}>
          <ToggleField
            {...register(`apiScopes.${index}.isPermitted` as const)}
            label={displayName ?? ""}
            className={css.input}
            defaultChecked={required}
            readOnly={required}
          />
          <Heading variant="h6" className={css.description}>
            {description}
          </Heading>
        </Fragment>
      )),
    [apiFields, register],
  );

  const handleClick = useCallback(
    (isPermitted: boolean) => async () => {
      const formData = getValues();

      formData.permissionGranted = isPermitted;

      try {
        const url = await mutateAsync(formData);

        window.location.href = url;
      } catch (e) {
        console.error(e);
      }
    },
    [getValues, mutateAsync],
  );

  return (
    <div
      className={classNames(css.ConsentForm, className)}
      data-testid="ConsentForm"
    >
      <Heading variant="h3" className={css.heading}>
        {clientName} is requesting your permissions
      </Heading>
      <Heading variant="h6" className={css.text}>
        Uncheck the permissions you do not wish to grant.
      </Heading>
      <form className={css.form}>
        <Fieldset className={css.fieldset} legend="Personal Information">
          {identityScopes}
        </Fieldset>
        <Fieldset className={css.fieldset} legend="Application Access">
          {apiScopes}
        </Fieldset>
        {allowRememberConsent && (
          <CheckboxField
            label="Remember my choice"
            className={css.input}
            {...register("rememberConsent")}
          />
        )}
        <div className={css.buttonWrapper}>
          <Button
            disabled={isLoading}
            loading={isLoading}
            onClick={handleClick(true)}
          >
            Allow
          </Button>
          <Button
            disabled={isLoading}
            loading={isLoading}
            onClick={handleClick(false)}
          >
            Deny
          </Button>
        </div>
      </form>
    </div>
  );
};

export { ConsentForm };

export default memo(ConsentForm);
