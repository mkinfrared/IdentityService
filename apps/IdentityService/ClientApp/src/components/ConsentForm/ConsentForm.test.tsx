import { fireEvent, render, waitFor } from "@testing-library/react";

import * as mutations from "api/mutations";
import { withQuery } from "utils/testUtils";

import { ConsentForm } from "./ConsentForm";

jest.mock("api/mutations");

describe("<ConsentForm />", () => {
  const mutationsMock = mutations as jest.Mocked<typeof mutations>;
  const returnUrl = "/marklar";
  const allowRememberConsent = true;
  const clientName = "marklar";

  const defaultValues = {
    rememberConsent: false,
    apiScopes: [
      {
        name: "OrdersApi",
        displayName: "OrdersApi",
        description: "",
        emphasize: false,
        required: false,
        isPermitted: false,
      },
    ],
    identityScopes: [
      {
        name: "openid",
        displayName: "Your user identifier",
        description: "",
        emphasize: false,
        required: true,
        isPermitted: true,
      },
      {
        name: "profile",
        displayName: "User profile",
        description:
          "Your user profile information (first name, last name, etc.)",
        emphasize: true,
        required: false,
        isPermitted: false,
      },
    ],
    description: "",
    permissionGranted: true,
    redirectUrl: "/foo/bar",
  };

  const Component = (
    <ConsentForm
      returnUrl={returnUrl}
      defaultValues={defaultValues}
      allowRememberConsent={allowRememberConsent}
      clientName={clientName}
    />
  );

  it("should be defined", () => {
    expect(ConsentForm).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component, { wrapper: withQuery });

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component, { wrapper: withQuery });
    const element = getByTestId("ConsentForm");

    expect(element).toBeDefined();
  });

  it("should call 'submitConsent' with 'permissionGranted' set to true", async () => {
    const data = {
      ...defaultValues,
      permissionGranted: true,
    };

    mutationsMock.submitConsent.mockResolvedValueOnce("marklar");

    const { queryByText } = render(Component, {
      wrapper: withQuery,
    });

    const allowButton = queryByText("Allow");

    fireEvent.click(allowButton!);

    await waitFor(() => {
      expect(mutationsMock.submitConsent).toHaveBeenCalled();

      expect(mutationsMock.submitConsent).toHaveBeenCalledTimes(1);

      expect(mutationsMock.submitConsent).toHaveBeenCalledWith(data);
    });
  });

  it("should call 'submitConsent' with 'permissionGranted' set to false", async () => {
    const data = {
      ...defaultValues,
      permissionGranted: true,
    };

    mutationsMock.submitConsent.mockResolvedValueOnce("marklar");

    const { queryByText } = render(Component, {
      wrapper: withQuery,
    });

    const allowButton = queryByText("Allow");

    fireEvent.click(allowButton!);

    await waitFor(() => {
      expect(mutationsMock.submitConsent).toHaveBeenCalled();

      expect(mutationsMock.submitConsent).toHaveBeenCalledTimes(1);

      expect(mutationsMock.submitConsent).toHaveBeenCalledWith(data);
    });
  });
});
