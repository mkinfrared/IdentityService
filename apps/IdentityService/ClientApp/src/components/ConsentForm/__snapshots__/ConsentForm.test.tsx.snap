// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConsentForm /> should match the snapshot 1`] = `
<div>
  <div
    class="ConsentForm"
    data-testid="ConsentForm"
  >
    <h3
      class="Heading-module_Heading__Wg8yB heading"
      data-testid="Heading"
    >
      marklar
       is requesting your permissions
    </h3>
    <h6
      class="Heading-module_Heading__Wg8yB text"
      data-testid="Heading"
    >
      Uncheck the permissions you do not wish to grant.
    </h6>
    <form
      class="form"
    >
      <fieldset
        class="Fieldset-module_Fieldset__2j-We fieldset"
        data-testid="Fieldset"
      >
        <legend
          class="Fieldset-module_legend__eWciB"
        >
          Personal Information
        </legend>
        <label
          class="Toggle-module_Toggle__l-SWp Toggle-module_readOnly__1ybfJ input"
          data-testid="Toggle"
        >
          <input
            checked=""
            class="Toggle-module_input__9vPu1"
            name="identityScopes.0.isPermitted"
            readonly=""
            type="checkbox"
          />
          <div
            class="Toggle-module_container__FoqlF"
          >
            <span
              class="Toggle-module_ball__bF3bD"
            />
          </div>
          <p
            class="Text-module_Text__K8mLZ Toggle-module_text__bqo8g"
            data-testid="Text"
          >
            Your user identifier
          </p>
        </label>
        <h6
          class="Heading-module_Heading__Wg8yB description"
          data-testid="Heading"
        />
        <label
          class="Toggle-module_Toggle__l-SWp input"
          data-testid="Toggle"
        >
          <input
            class="Toggle-module_input__9vPu1"
            name="identityScopes.1.isPermitted"
            type="checkbox"
          />
          <div
            class="Toggle-module_container__FoqlF"
          >
            <span
              class="Toggle-module_ball__bF3bD"
            />
          </div>
          <p
            class="Text-module_Text__K8mLZ Toggle-module_text__bqo8g"
            data-testid="Text"
          >
            User profile
          </p>
        </label>
        <h6
          class="Heading-module_Heading__Wg8yB description"
          data-testid="Heading"
        >
          Your user profile information (first name, last name, etc.)
        </h6>
      </fieldset>
      <fieldset
        class="Fieldset-module_Fieldset__2j-We fieldset"
        data-testid="Fieldset"
      >
        <legend
          class="Fieldset-module_legend__eWciB"
        >
          Application Access
        </legend>
        <label
          class="Toggle-module_Toggle__l-SWp input"
          data-testid="Toggle"
        >
          <input
            class="Toggle-module_input__9vPu1"
            name="apiScopes.0.isPermitted"
            type="checkbox"
          />
          <div
            class="Toggle-module_container__FoqlF"
          >
            <span
              class="Toggle-module_ball__bF3bD"
            />
          </div>
          <p
            class="Text-module_Text__K8mLZ Toggle-module_text__bqo8g"
            data-testid="Text"
          >
            OrdersApi
          </p>
        </label>
        <h6
          class="Heading-module_Heading__Wg8yB description"
          data-testid="Heading"
        />
      </fieldset>
      <label
        class="Checkbox-module_Checkbox__ChGlN input"
        data-testid="Checkbox"
      >
        <input
          class="Checkbox-module_input__-039s"
          name="rememberConsent"
          type="checkbox"
          value=""
        />
        <div
          class="Checkbox-module_checkmark__pjJT2"
        >
          <svg
            class="Checkbox-module_tick__-Yp8y"
            viewBox="0 0 1280 1186"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1253 5.4c-3 1.8-14.5 8.6-25.5 15.1C1114 87.6 990.7 180.8 869 291.4c-64.7 58.9-140.3 134.2-203 202.1-108.5 117.6-218 252.3-311.3 382.9l-4.8 6.7-3.2-1.9c-1.8-1.1-75.8-43-164.6-93.2-130.9-73.9-161.6-90.9-162.6-89.8-6.2 6.8-15.6 19.1-15.3 20 .2.6 90.8 106.1 201.4 234.4l200.9 233.3h7.5l7.6.1 15.3-30.8c174.3-350.6 357.9-626.1 584-876.7 66-73.1 162.4-170.1 238.1-239.4 8.5-7.8 15.4-14.8 15.4-15.6-.2-1.5-14.5-21.4-15.4-21.4-.3 0-3 1.4-6 3.3z"
            />
          </svg>
        </div>
        <p
          class="Text-module_Text__K8mLZ Checkbox-module_text__usAwV"
          data-testid="Text"
        >
          Remember my choice
        </p>
      </label>
      <div
        class="buttonWrapper"
      >
        <button
          class="Button-module_Button__0OJEu Button-module_opaque__O0DEF"
          data-testid="Button"
          type="button"
        >
          <svg
            class=""
            fill="#fff"
            viewBox="0 0 135 140"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="120"
              rx="6"
              width="15"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="30"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="140"
              rx="6"
              width="15"
              x="60"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="90"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="120"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
          </svg>
          Allow
        </button>
        <button
          class="Button-module_Button__0OJEu Button-module_opaque__O0DEF"
          data-testid="Button"
          type="button"
        >
          <svg
            class=""
            fill="#fff"
            viewBox="0 0 135 140"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="120"
              rx="6"
              width="15"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="30"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="140"
              rx="6"
              width="15"
              x="60"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="90"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="120"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
          </svg>
          Deny
        </button>
      </div>
    </form>
  </div>
</div>
`;
